kt/godot-library/src/main/kotlin/godot/core/memory/MemoryManager.kt

```diff
@@ -293,95 +199,44 @@ internal object MemoryManager {
-        executor.awaitTermination(5000, TimeUnit.MILLISECONDS)
-    }
-
-    @Suppress("unused")
-    fun cleanUp() {
-        for (singletonIndex in singletonIndexes) {
-            val id = ObjectDB[singletonIndex]!!.id
-            unbindInstance(id.id)
-            ObjectDB[singletonIndex] = null
+    fun preCleanup() {
+        for (callback in cleanupCallbacks) {
+            callback.invoke()
         }
 
-        while (staticInstances.size > 0) {
-            val iterator = staticInstances.iterator()
-            staticInstances = mutableSetOf()
-            for (instance in iterator) {
-                instance.collect()
-            }
+        // Get through all remaining [Object] and remove them from the ObjectDB. It will remove all singletons.
+        for (ref in ObjectDB.filterNotNull().filter{!it.objectID.isReference}) {
+            ObjectDB[ref.objectID.index] = null
         }
+    }
 
-        // Clear any cached StringName or NodePath objects.
-        stringNameCache.clear()
-        nodePathCache.clear()
-
-        var begin = Instant.now()
-        while (ObjectDB.any { it != null } || nativeCoreTypeMap.isNotEmpty()) {
-
-            forceJvmGc()
-            if (manageMemory()) {
-                begin = Instant.now()
-            }
-
-            val finish = Instant.now()
-            if (Duration.between(begin, finish).toMillis() > 5000) {
-                warning(
-                    buildString {
-                        appendLine("Some JVM godot instances are leaked.")
-                        if (shouldDisplayLeakInstancesOnClose) {
-                            val leakedObjects = ObjectDB.filterNotNull()
-                            appendLine("${leakedObjects.size} Objects:")
-                            for (entry in leakedObjects) {
-                                val obj = entry.get()!!.value!!
-                                append("    ${obj::class.simpleName} ${entry.id.id} ")
-                                append("C++ instance alive: ${checkInstance(obj.rawPtr, obj.id.id)}")
-                                append(System.lineSeparator())
-                            }
-                            appendLine("${nativeCoreTypeMap.size} Leaked native core types:")
-                            for (entry in nativeCoreTypeMap) {
-                                append("    ${entry.key}: ${entry.value.get()}")
-                                append(System.lineSeparator())
-                            }
-                        }
-                    }
-                )
-                notifyLeak()
-                break
+    fun checkCleanUp(): Boolean {
+        if (ObjectDB.any { it != null } && nativeCoreTypeMap.isNotEmpty()) {
+            // Still work to do so we force the gc
+            var any: Any? = Any()
+            val wkRef = WeakReference(any)
+            @Suppress("UNUSED_VALUE")
+            any = null
+            while (wkRef.get() != null) {
+                System.gc()
```

Comment:
Didn't we want to get rid of this?

Reply:
I'm keeping it for the debugging phase right now. Depending on how it goes, I might have to keep it, which I would prefer not.